src/cascadia/TerminalApp/SuggestionsControl.cpp

```diff
@@ -0,0 +1,1103 @@
+            }
+
+            if (_filteredActions.GetAt(i).Item() != actions[i].Item())
+            {
+                _filteredActions.InsertAt(i, actions[i]);
+            }
+        }
+
+        // Remove any extra trailing items from the destination
+        while (_filteredActions.Size() > actions.size())
+        {
+            _filteredActions.RemoveAtEnd();
+        }
+
+        // Add any extra trailing items from the source
+        while (_filteredActions.Size() < actions.size())
+        {
+            _filteredActions.Append(actions[_filteredActions.Size()]);
+        }
+    }
+
+    // Method Description:
+    // - Update the list of current nested commands to match that of the
+    //   given parent command.
+    // Arguments:
+    // - parentCommand: the command with an optional list of nested commands.
+    // Return Value:
+    // - <none>
+    void SuggestionsControl::_updateCurrentNestedCommands(const winrt::Microsoft::Terminal::Settings::Model::Command& parentCommand)
+    {
+        _currentNestedCommands.Clear();
+        for (const auto& nameAndCommand : parentCommand.NestedCommands())
+        {
+            const auto action = nameAndCommand.Value();
+            auto nestedActionPaletteItem{ winrt::make<winrt::TerminalApp::implementation::ActionPaletteItem>(action) };
+            auto nestedFilteredCommand{ winrt::make<FilteredCommand>(nestedActionPaletteItem) };
+            _currentNestedCommands.Append(nestedFilteredCommand);
+        }
+    }
+
+    // Method Description:
+    // - Dismiss the command palette. This will:
+    //   * select all the current text in the input box
+    //   * set our visibility to Hidden
+    //   * raise our Closed event, so the page can return focus to the active Terminal
+    // Arguments:
+    // - <none>
+    // Return Value:
+    // - <none>
+    void SuggestionsControl::_close()
+    {
+        Visibility(Visibility::Collapsed);
+
+        // Clear the text box each time we close the dialog. This is consistent with VsCode.
+        _searchBox().Text(L"");
+
+        _nestedActionStack.Clear();
+
+        ParentCommandName(L"");
+        _currentNestedCommands.Clear();
+
+        PreviewAction.raise(*this, nullptr);
+    }
+
+    // Method Description:
+    // - This event is triggered when filteredActionView is looking for item container (ListViewItem)
+    // to use to present the filtered actions.
+    // GH#9288: unfortunately the default lookup seems to choose items with wrong data templates,
+    // e.g., using DataTemplate rendering actions for tab palette items.
+    // We handle this event by manually selecting an item from the cache.
+    // If no item is found we allocate a new one.
+    // Arguments:
+    // - args: the ChoosingItemContainerEventArgs allowing to get container candidate suggested by the
+    // system and replace it with another candidate if required.
+    // Return Value:
+    // - <none>
+    void SuggestionsControl::_choosingItemContainer(
```

Comment:
we don't actually have different Items in this view. Will we ever? Is this future-cruft?

Reply:
This seemed scary like it was load bearing that I didn't touch it. Let's see how load bearing it really is...

Reply:
Oh I think we do use it for `GeneralItemTemplate` vs `NestedItemTemplate`

Reply:
ah gotcha